src/audio_unit/looper.rs

```diff
@@ -0,0 +1,137 @@
+use crate::{audio_unit::AudioUnit, util, Result};
+use cpal::StreamConfig;
+use std::sync::mpsc::{self, Receiver, Sender};
+use State::*;
+
+#[derive(Debug)]
+pub enum Message {
+    Toggle,
+}
+
+#[derive(Debug)]
+enum State {
+    Off,
+    Recording { position: usize },
```

Comment:
What really is a `usize`? Is it always an "index" into eg an array and if so why is that a special type? I feel like I've read that it's equivalent to a pointer type (ie it can fit a memory address), but indexes aren't pointers?

Reply:
I still don't get why an index == a pointer, one is a small integer (that doesn't eg have to get directly added to a pointer address in pointer arithmetic - I guess indexes would get _multiplied_ by a pointer size in pointer arithmetic, but does that imply it needs to be the same type as the thing it's being multiplied by?) and the other is a memory address

Reply:
I think it would need to be casted/converted to the same type as the pointer to do math with it (at least to do that math efficiently).

It does feel a little leaky in a sense, but I'm not sure what the alternative would be. I think the key is that "one is a small integer" is not true.

On a 32-bit CPU, the maximum length of a `Vec` is (handwave, handwave) 2^32 (a `Vec` of `bool`s, which consumes the entire addressable memory), so the index type needs to be >= `u32`.

On a 64-bit CPU, the maximum length of a `Vec` is 2^64, so the index type needs to be >= `u64`.

So the options are:

- `u32` on all architectures:
  - good on 32-bit
  - arbitrarily restricts the length of `Vec`s on 64-bit
- `u64` on all architectures:
  - works on 32-bit (I think), but with a significant performance penality when doing math, since it's not a native type to the CPU. also, it's impossible to use the top 32-bits of the index because your program would always crash when trying to allocate an array longer than 2^32
  - good on 64-bit
- arbitrary precision integer type
  - worst of both previous choices: *way* more expensive than native types on all architectures
  - no gains (the amount of addressable memory doesn't change)
- `usize`
  - allows full usage of the addressable memory on both architectures
  - efficient on both architectures since it will always be a "native" type

Reply:
arbitrary precision: "no gains" other than consistency, I guess